Expose get_build_command() as requested in issue #8511

behlec opened the pull request:
Expose `get_build_command()` in meson as requested in #8511

First draft. I am not sure if the return type is as intended. Currently it is the same as of the internal function `get_build_command()`.

xclaesse commented:
It seems a bit too niche to have a top level function IMHO. What about making this a method on the meson object? `meson.get_command()` ?

xclaesse commented on mesonbuild/interpreter.py:
```diff
@@ -1911,6 +1911,7 @@ def __init__(self, build: 'build.Build', interpreter: 'Interpreter'):
                              'has_external_property': self.has_external_property_method,
                              'backend': self.backend_method,
                              'add_devenv': self.add_devenv_method,
+                             'get_command': self.get_command,
```
The convention is to have _method suffix.

xclaesse commented on mesonbuild/interpreter.py:
```diff
@@ -2232,6 +2233,14 @@ def add_devenv_method(self, args: T.Union[str, list, dict, EnvironmentVariablesH
+    @noPosargs
+    @permittedKwargs({'unbuffered'})
+    def get_command(self, args, kwargs):
+        unbuffered = kwargs.get('check', False)
```
```suggestion
        unbuffered = kwargs.get('unbuffered', False)
```

I'm not sure we need to expose that, what does it do?

behlec replied:
It adds the `-u` flag to the call. I.e. `[ '/usr/bin/python3', '/foo/meson.py' ]` becomes `['/usr/bin/python3', '-u', '/foo/meson.py']`. Not sure if that is necessary. I included it, since the original method provides it. I am happy to remove that part.

xclaesse reviewed (commented):
Just a couple nitpickings in the implementation. To be complet, would be nice to add:
- reference manual doc
- release notes snippet
- unit test

xclaesse commented on mesonbuild/interpreter.py:
```diff
@@ -2232,6 +2233,14 @@ def add_devenv_method(self, args: T.Union[str, list, dict, EnvironmentVariablesH
 
+    @noPosargs
+    @permittedKwargs({'unbuffered'})
+    def get_command(self, args, kwargs):
```
Missing @FeatureNew('meson.get_command', '0.58.0')

behlec replied:
I will add the requested parts if we are clear on whether we want to support unbuffered or not.
> * reference manual doc
> * release notes snippet
> * unit test

For the final pull request I will squash the branch, this is a small, on commit change.

behlec commented:
I have removed the `unbuffered` kwarg. Added documentation and a simple test.
Trying to write the test I stumbled across the problem, that I cannot compare the output of `run_command.stdout()` with `meson.version()`, since they have different types.

xclaesse commented on mesonbuild/interpreter.py:
```diff
@@ -2221,6 +2222,11 @@ def add_devenv_method(self, args: T.Union[str, list, dict, EnvironmentVariablesH
 
+    @FeatureNew('meson.get_command', '0.58.0')
+    @noPosargs
+    def get_command_method(self, args, kwargs):
```
`@noKwargs` too.

xclaesse reviewed (commented):
> Trying to write the test I stumbled across the problem, that I cannot compare the output of `run_command.stdout()` with `meson.version()`, since they have different types.

Hmmm, that seems to be a regression in the interpreter that I made when adding some special casing for version string.
```
class MesonVersionString(str):
    pass
```

As a work around for now, that works: `assert('0.57.999' == '@0@'.format(meson.version()))`.

xclaesse commented:
Looks good to me, could you squash please? Thanks.

jpakkane commented:
This is slightly different than what I got from the original bug report. If we expose something like 'MESONCOMMAND` in an env var like `MESONINTROSPECT` then would that provide for everything needed? It seems that the most common usage of this would be to pass it to external scripts so they can invoke Meson. Putting it in a common envvar would make usage simpler.

behlec commented:
I see what you mean. I will start working on that. However, maybe it would be a good idea if OP could comment, what he had exactly in mind?
Until we are sure, I will keep the current commit and add MESONCOMMAND to envar in a separate commit.

behlec commented:
First draft. If I got the idea right I will add test, documentation, and update the release snippet.

behlec commented:
So according to the requester of the issue he wants an `meson.build` command that returns an array.

behlec commented:
I have rebased the changes on master. For now I have only added `meson.get_command()`. This is what the requester had in mind. If you want I will also add `MESONCOMMAND` again.

anarazel commented:
> This is slightly different than what I got from the original bug report. If we expose something like 'MESONCOMMAND`in an env var like`MESONINTROSPECT` then would that provide for everything needed? It seems that the most common usage of this would be to pass it to external scripts so they can invoke Meson. Putting it in a common envvar would make usage simpler.

FWIW, I was thinking of a function like meson.get_build_command() that returns an array (like `['path/to/python', '/path/to/meson.py']`). Although `MESONCOMMAND` could also be useful. The advantage of the function would be that it's trivial to write a run_command() executing meson - whereas  `MESONCOMMAND` wouldn't be as easy to execute in a portable fashion.

anarazel commented:
@behlec  This is great - sorry that I missed you opening this PR.

The only substantive comment about the changes is that it might make sense to change the function name. The compiler objects have `cmd_array(): returns an array containing the command(s) for the compiler.`, so maybe it'd make sense to name this `meson.meson_cmd_array()`?

behlec commented:
I have renamed the command as wished.
What version will this be in? I have set it to 0.60.0 for now.

As a reminder, I do have patches to add the environment variables if wished. Let me know if I should add them.

xclaesse commented:
Saying that because in report https://github.com/mesonbuild/meson/issues/8511 @anarazel's first reflex was to try that, so presumably it's the most discoverable way of doing it?

eli-schwartz commented:
That sounds like an excellent idea, much preferred.

behlec commented:
Let me know what the final decision is and I will make the changes.

xclaesse commented:
In interpreter.program_lookup() we already special case 'python3', IMHO it makes sense to handle 'meson' there the same way.

anarazel commented:
@xclaesse 
> I just had an idea: what about instead special case `find_program('meson')` ?
> Saying that because in report #8511 @anarazel's first reflex was to try that, so presumably it's the most discoverable way of doing it?

It'd work for my current needs. I'm mildly worried that there could be cases where the fact that the path will likely consist out of two components would cause problems. E.g. because it's embedded in a generated script or such.

behlec commented:
> In interpreter.program_lookup() we already special case 'python3', IMHO it makes sense to handle 'meson' there the same way.

So `find_program('meson')` returns always the current meson location, even if it is not in PATH?
If I read the case for python3 correctly, meson searches for python3 and only returns the currently used location, if it cannot find python3 in the regular way.
@anzarel mentioned, he could not use it, because  the meson instance used might not be in PATH.

If we go the `find_program` route we need to carefully decide how to handle all the options. What happens if native is set to false? What happens if override_find_program is set?

anarazel commented:
I personally see the `meson_cmd_array()` approach as superior (better multi-component handling, less magic). While @xclaesse is right that I looked at find_program() for a solution, I don't think I would have if `meson_cmd_array()` existed...

eli-schwartz commented:
@behlec we could raise an error message in both cases.

@anarazel Hmm, how so? find_program already sometimes is an array command, e.g. smart shebang handling, required interpreters for scripts on windows where you can't run anything from cmd, that isn't an executable


It is true that this would be a bit magic. :( Ok the other hand, it folds into existing API rather than adding new API. Dunno which is better.

eli-schwartz commented:
Also extending find_program means that people currently using
```meson
run_target('meson', ...)
```

Will be automatically upgraded.

xclaesse commented:
I have to admit, I'm undecided, both solutions have pros/cons. We already have `compiler.cmd_array()` so it's try that it's not a new type of API neither. One thing is I don't like the repetition of "meson" in `meson.meson_cmd_array()`, but I think that was done on purpose to avoid confusion.

@jpakkane What's your opinion? `find_program('meson')`, `meson.cmd_array()` or `meson.meson_cmd_array()`?

Akaricchi commented:
I think `find_program('meson')` on its own is a bad idea. The external program object does not have a method to return the command as an array of args, it only has `full_path()` which returns a string. It's not clear how to handle commands with args there. Extending `find_program` for consistency and use cases like `run_target('meson', ...)` still seems worthwhile, however.

We can add `cmd_array()` to the external program object, and then we don't need `meson.cmd_array()`. This may be the best solution overall: most general and consistent with the compiler object.

tristan957 commented:
@jpakkane seems like this PR stalled pending your input.

tristan957 commented:
> I think `find_program('meson')` on its own is a bad idea. The external program object does not have a method to return the command as an array of args, it only has `full_path()` which returns a string. It's not clear how to handle commands with args there. Extending `find_program` for consistency and use cases like `run_target('meson', ...)` still seems worthwhile, however.
> 
> We can add `cmd_array()` to the external program object, and then we don't need `meson.cmd_array()`. This may be the best solution overall: most general and consistent with the compiler object.

I like this idea quite a bit. Is anyone against it?

anarazel commented:
> > We can add `cmd_array()` to the external program object, and then we don't need `meson.cmd_array()`. This may be the best solution overall: most general and consistent with the compiler object.
> 
> I like this idea quite a bit. Is anyone against it?

That seems like a good idea, regardless of the this issue's resolution.


> I think find_program('meson') on its own is a bad idea. The external program object does not have a method to return the command as an array of args, it only has full_path() which returns a string. It's not clear how to handle commands with args there. Extending find_program for consistency and use cases like run_target('meson', ...) still seems worthwhile, however.

I don't love magic in  `find_program()`, it feels hard to discover. What about `meson.get_meson_program()` or such? The proposed `Program.cmd_array()` could then be used to execute commands?

xclaesse commented:
What I like with `find_program('meson')` is it will automatically improve existing code that already do it, or do `run_target('meson', ...)`. IIRC custom_target(), run_target(), etc, already pass argv[0] through find_program() and that will already work when `meson` is in PATH. Adding the special casing will make it work in more situations. The best feature is a feature you don't even have to discover because it just works.

xclaesse commented:
> The external program object does not have a method to return the command as an array of args, it only has `full_path()` which returns a string. It's not clear how to handle commands with args there.

Why would you need the command array at all? Meson will already automatically expand any program returned by `find_program()` to a string array when used inside a command arguments, even when it's not used as argv[0]. I'm not sure what's the use-case here.

tristan957 commented:
I decided to put up a PR for `find_program('meson')` at https://github.com/mesonbuild/meson/pull/11849.

tristan957 commented:
@Akaricchi I think you could just switch to that now and get the same behavior as you have now, but also get the upgraded behavior when the PR merges.

xclaesse commented:
> You really should be asking why does `.full_path()` need to exist, because that's essentially a broken version of the proposed `.cmd_array()`.

That's a fair point, `.full_path()` should probably be deprecated, question is does it need to be replaced by `.cmd_array()`.

tristan957 commented:
I don't really understand why it would need to be deprecated.

xclaesse commented:
Let's close, we have find_program('meson') now.
